### ISSUE TITLE
Access token validation issue

### ISSUE THREAD

**Issue by MikiStojakovic** [IdentityServer/IdentityServer4.Samples#206]
Hi @leastprivilege @brockallen @scottbrady91 

I am trying in last few days to figure out how to set up my resource server to consume access token.
I strive to learn how to use Identity Server 4 and I don't have idea how to solve this.

I have ang2 app which call Identity Server separate solution and I get id_token as well as access_token by using implicit flow. I add access_token to authorization header of ang2 app and call resource api. Although I add IdentityServer middleware to api StartUp class I still getting 401.

Just to point out it is AspNet Core version 1.1.2 of resource api.

So here in browser I see authorization header is present with access_token in it.
![ang_request](https://user-images.githubusercontent.com/17271707/30397460-48acacaa-98cd-11e7-8104-850cf3d6bd9e.png)

Here is my Startup class on resource api side:
public class Startup
    {
        public Startup(IHostingEnvironment env)
        {
            var builder = new ConfigurationBuilder()
                .SetBasePath(env.ContentRootPath)
                .AddJsonFile("appsettings.json", optional: false, reloadOnChange: true)
                .AddJsonFile($"appsettings.{env.EnvironmentName}.json", optional: true)
                .AddEnvironmentVariables();
            Configuration = builder.Build();
        }

        public IConfigurationRoot Configuration { get; }

        // This method gets called by the runtime. Use this method to add services to the container.
        public void ConfigureServices(IServiceCollection services)
        {
            // IMPORTANT for Angular CLI Server that runs separate Web site
            services.AddCors(options =>
            {                
                options.AddPolicy("CorsPolicy",builder => 
                    builder
                        .AllowAnyOrigin()
                        .AllowAnyMethod()
                        .AllowAnyHeader()
                        .AllowCredentials() );
            }); 
            
            // Add framework services.
            services.AddMvc();
        }

        // This method gets called by the runtime. Use this method to configure the HTTP request pipeline.
        public void Configure(IApplicationBuilder app, IHostingEnvironment env, ILoggerFactory loggerFactory)
        {
            loggerFactory.AddConsole(Configuration.GetSection("Logging"));
            loggerFactory.AddDebug();

            loggerFactory.AddConsole(LogLevel.Trace);
            app.UseCors("CorsPolicy");
            
             app.UseIdentityServerAuthentication(new IdentityServerAuthenticationOptions
            {
                AuthenticationScheme = "Bearer",
                Authority = "http://localhost:5000",
                RequireHttpsMetadata = false,

                ApiName = "todo"               
            });

            app.UseMvc();
        }
    }

Here is the logs on api side:
![api_log_error_1](https://user-images.githubusercontent.com/17271707/30397643-d363cd92-98cd-11e7-8a25-869bb44f9228.png)
![api_log_error_2](https://user-images.githubusercontent.com/17271707/30397647-d905afea-98cd-11e7-8d97-a92bbd36f6bb.png)

Am I doing something wrong? Please can you help me with this?

**Comment by brockallen** [comment]
Any update?

**Comment by MikiStojakovic** [comment]
Hi  @brockallen

I was looking your course with Scott Hanselman on MVA and based on that I made some changes into my asp net core api middleware. Now I am using JwtBearerAuthentication middleware as you can see on following picture:
![api_log](https://user-images.githubusercontent.com/17271707/31054339-05dbdada-a6b1-11e7-8c10-77846dba34cc.png)
Now I get "No SecurityTokenValidator available for token as it is shown bellow:
![api_log](https://user-images.githubusercontent.com/17271707/31054360-46d923f8-a6b1-11e7-9c2a-35c07b90105d.png)
The only difference in comparison to your MVA course example is that I use AddDeveloperSinginCredential so I don't use any certificate to sign token, bellow is id4 middleware (pic bellow)
![id4_side](https://user-images.githubusercontent.com/17271707/31054387-e851b628-a6b1-11e7-890f-c5c7772873b1.png)

I made question regarding this on stackoverflow, here is the link
https://stackoverflow.com/questions/46441589/identity-server-4-no-security-token-validator-available-for-token

Please, can you give some hint on this?
Thanks in advance.

**Comment by brockallen** [comment]
I'm pretty sure I did not show the settings to disable validating issuer and the signature for the JWT :)

> The only difference in comparison to your MVA course example is that I use AddDeveloperSinginCredential so I don't use any certificate to sign token, bellow is id4 middleware (pic bellow)

You still are using keys to sign the token -- that's always still there.

I can't tell from the above what the issue is, sorry.

**Comment by MikiStojakovic** [comment]
Thanks on your answer and you are right about 

> I did not show the settings to disable validating issuer and the signature for the JWT :)

When I said that I changed code regarding MVA course I was thinking on api middleware, not on tokenOption. That was just my try to see can I make it work. Anyhow I removed tokenOption where I disabled validating issuer and the signature for the JWT and I get the same error again.

Then it just dawn on me that I need somehow to make my api to use tempkey.rsa which is generated by Id server app. I don't know how to do that even I was googling. I didn't find any understandable example and make api be able to verify signature of access token.  I assume it is the last part of the puzzle I need.

Can you provide any link, hint, anything on how to implement and use tempkey.rsa in api solution?

P.S. Course on VMA is very good. I hope it will be more courses to help us to learn faster.

Thanks in advance.

**Comment by brockallen** [comment]
> Then it just dawn on me that I need somehow to make my api to use tempkey.rsa which is generated by Id server app. I don't know how to do that even I was googling.

It does it automatically because you have set the `Authority` -- it fetches it from the discovery document.

**Comment by MikiStojakovic** [comment]
I solved this at the end. I updated asp.net core to version 2 and added new AspNetCore Authentication JwtBearer middleware in Startup class of Api project and now everything works fine.

I will attach Startup class of my api project, maybe it can help to someone.
![api_startup_class_conf_service_method](https://user-images.githubusercontent.com/17271707/31575646-de1ab846-b0eb-11e7-8c5f-9478a0ed58c2.png)
![api_startup_class_configure_method](https://user-images.githubusercontent.com/17271707/31575649-e2c51990-b0eb-11e7-93e3-9ae474060faf.png)
